internal/objectstore/s3objectstore.go

```diff
@@ -0,0 +1,383 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package objectstore
+
+import (
+	"context"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/core/objectstore"
+)
+
+const (
+	defaultBucketName = "juju"
+)
+
+type s3ObjectStore struct {
+	baseObjectStore
+	client    objectstore.Client
+	namespace string
+	requests  chan request
+}
+
+// NewS3ObjectStore returns a new object store worker based on the s3 backing
+// storage.
+func NewS3ObjectStore(ctx context.Context, namespace string, client objectstore.Client, metadataService objectstore.ObjectStoreMetadata, claimer Claimer, logger Logger, clock clock.Clock) (TrackedObjectStore, error) {
```

Comment:
Worth creating a struct for these args (not including context of course)?

Reply:
It's getting that way!